File: csharp/collatz-conjecture/CollatzConjecture.cs
public static class CollatzConjecture
{
    public static int Steps(int number)
    {
        if (number < 1)
            throw new System.ArgumentException("number must be strictly positive and not null");
        var counter = 0;
        long longNumber = number;
        return Step(longNumber, counter);
    }


    private static int Step(long number, int counter)
    {
        long newNumber = number % 2 == 0 ? number / 2 : 3 * number + 1;
        return number == 1 ? counter : Step(newNumber, ++counter);
    }

}

File: python/collatz-conjecture/collatz_conjecture.py
def steps(number):
    if number <= 0:
        raise ValueError("number must be strictly posisitve")
    return count_steps(number, 0)

def count_steps(number, step = 0):
    if number == 1:
        return step
    else:
        step += 1
        number = number / 2 if number % 2 == 0 else 3 * number + 1
        return count_steps(number, step)
